addons/analytic/models/analytic_plan.py

```diff
@@ -244,38 +245,71 @@ def _get_applicability(self, **kwargs):
 
     def unlink(self):
         # Remove the dynamic field created with the plan (see `_inverse_name`)
+        if not self:
+            return super().unlink()
         self._find_plan_column().unlink()
         return super().unlink()
 
-    def _find_plan_column(self):
-        return self.env['ir.model.fields'].sudo().search([
-            ('name', 'in', [plan._strict_column_name() for plan in self]),
-            ('model', '=', 'account.analytic.line'),
-        ])
+    def _get_plan_column_models(self):
+        # Return the models on which the plan column should be created
+        return ['account.analytic.line']
+
+    def _find_plan_column(self, model=False):
+        domain = [('name', '=', self._strict_column_name())]
+        if model:
+            domain.append(('model', '=', model))
+        return self.env['ir.model.fields'].search(domain)
 
     def _sync_plan_column(self):
-        # Create/delete a new field/column on analytic lines for this plan, and keep the name in sync.
-        for plan in self:
-            prev = plan._find_plan_column()
-            if plan.parent_id and prev:
-                prev.unlink()
-            elif prev:
-                prev.field_description = plan.name
-            elif not plan.parent_id:
-                column = plan._strict_column_name()
-                self.env['ir.model.fields'].with_context(update_custom_fields=True).sudo().create({
-                    'name': column,
-                    'field_description': plan.name,
-                    'state': 'manual',
-                    'model': 'account.analytic.line',
-                    'model_id': self.env['ir.model']._get_id('account.analytic.line'),
-                    'ttype': 'many2one',
-                    'relation': 'account.analytic.account',
-                    'store': True,
-                })
-                tablename = self.env['account.analytic.line']._table
-                indexname = make_index_name(tablename, column)
-                create_index(self.env.cr, indexname, tablename, [column], 'btree', f'{column} IS NOT NULL')
+        # Create/delete a new field/column on related models for this plan, and keep the name in sync.
+        models = self._get_plan_column_models()
+        for model in models:
```

Comment:
Don't shadow `models` imported at top level (defining the variable is not really useful anyway)